notes/radicle-p2p-hypercore-overview.md

```diff
@@ -0,0 +1,114 @@
+and replicate the feed.
+
+We say a user *hosts a feed* if the user makes it available to the world. If a
+users hosts a feed the user is either the owner of that feed or she replicates
+it.
+
+`hypercore` supports feed subscriptions. A user can subscribe to a feed and will
+be notified if the owner publishes a new entry.
+
+Every feed a user owns or replicates is stored locally and updated automatically
+if a user goes online.
+
+`discovery-swarm` is a library to discover hosts for hypercore feeds and
+announce the hosting of a feed. It uses the Bittorrent DHT and centralised DNS
+for global discovery as well as Multicast DNS for local discovery.
+
+
+Single-writer Chains
+--------------------
+
+*Single-writer chains* are Radicle chains that only the owner of a specific
+private key can write to.
+
+We use hypercore feeds to represent single-writer chains. Each entry in the feed
+is a Radicle expression. Single-writer chains are identified by the discovery
+key of their feed.
+
+To read a Radicle chain a user needs the discovery key of a feed. With the
+discovery key the user uses hyperdiscovery to find someone who replicates the
+feed. (This may be the owner.) The user then reads all entries in the feed with
+hypercore and evaluates them with radicle.
+
+To write an expression to a Radicle chain a user publishes a new entry to the
+`hypercore` feed.
+
+
+Multi-writer Chains
+-------------------
+
+*Multi-writer chains* are Radicle chains that the owners of multiple distinct
+private keys can write to.
+
+We implement a multi-writer chain by using set of feeds owned by different
+users. Every entry in these feeds is a Radicle expression and a link to another
+entry in any of feeds. Every entry in a feed now determines a linked list of
+Radicle expression (assuming the feed entries and links are well formed and
+valid) and thus a Radicle chain.
+
+To derive a Radicle chain from a set of feeds we need an algorithm that takes
+the heads of all the feeds of a multi-writer chain and outputs either a chain of
+Radicle expressions or signals that the chain is invalid.
+
+The simplest such algorithm is the following: We accept only the situation where
+there is a feed with head entry `e` so that the head entries of all other feeds
+are ancestors of `e`. In that case `e` is the chain derived from the set of
+feeds. Otherwise the chain is invalid.
+
+
+**TODO** Properties of this scheme
+
+**TODO** Addressing entries
+
+
+Offline First and Replication
+-----------------------------
+
+**TODO** flesh this out
+
+Hypercore stores every feed a user is interested locally.
+
+Hypercore allows one user to replicate the feed of another user. Updates are
+propagated automatically.
+
+Users can fetch a feed from a user replicating it if the owner is not online.
+
+
+Security
+--------
+
+* Feeds can be censored. If the owner goes offline and all users that replicate
+  the feed conspire to drop part of the head of the feed. If one user chooses to
```

Comment:
> drop part of the head

Drop the head? 

Is that an actual concern? Replicators can conspire to not propagate the latest update.

Reply:
> Is that an actual concern? Replicators can conspire to not propagate the latest update.

I don’t think it is an actual concern. I just wanted to make a note of it.